### PR TITLE
Implement transpose in `DeviceInstance::HostBufferToDevice`

### PR DESCRIPTION
**Pull request by ramiro050**
This patch adds support for input tensors to `HostBufferToDevice` that are not dense-row-major. This is achieved by transposing the input data on the host before moving it to the device.

### CODE REVIEW

File: iree/integrations/pjrt/common/api_impl.cc [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include "iree/hal/api.h"
 #include "iree/integrations/pjrt/common/iree_helpers.h"
 #include "iree/integrations/pjrt/common/tensor_utils.h"
+#include "xla/pjrt/transpose.h"
```
Comment (ramiro050) [inline iree/integrations/pjrt/common/api_impl.cc:15]: The current implementation works. The only thing missing is moving the header file `transpose.h` into `openxla-pjrt-plugin`.

Comment (jpienaar) [reply iree/integrations/pjrt/common/api_impl.cc:15]: Do we need to move it? We have a dep on that repo already.

Comment (ramiro050) [reply iree/integrations/pjrt/common/api_impl.cc:15]: I wasn't sure what the rules were regarding depending on xla files. If this is fine, then this is ready for review!

Comment (jpienaar) [reply iree/integrations/pjrt/common/api_impl.cc:15]: Lets start like this and if we notice pains, we copy :) (I think we want to consider the codegen option that Rob had mentioned, so starting with this is good).

**Comment by jpienaar** [comment]
Could you add a test? Something akin to what was in the issue along with the smoke test. We may want to emit a warning in debug stream saying that we are incurring a cost (2x memory on host and runtime cost) so that when we see this in a profile later, its not unexpected. It is avoidable doing it here by folding program side, so might be good having a feature request for that that we could link from here.

**Comment by ramiro050** [comment]
I've added a test and a warning with a link to the feature request. There are some JAX tests that still don't pass, so I'm currently investigating those.

**Comment by rsuderman** [comment]
> I've added a test and a warning with a link to the feature request. There are some JAX tests that still don't pass, so I'm currently investigating those.

How many pass vs fail? Most just happen to use the transposed memory behavior and may also test other behavior.

**Comment by ramiro050** [comment]
> > I've added a test and a warning with a link to the feature request. There are some JAX tests that still don't pass, so I'm currently investigating those.
> 
> How many pass vs fail? Most just happen to use the transposed memory behavior and may also test other behavior.

Haven't been able to run the full test suite because of crashes (pytest AFAIK is not smart enough to keep going after a segfault...), but so far, 118/214 pass and 3/214 crash. Some of the failures do seem related to the `TransposePlan` creation, so I'm looking into those.

I'll update with full test suite numbers tomorrow.

**Comment by ramiro050** [comment]
> pytest AFAIK is not smart enough to keep going after a segfault...

`pytest-xdist` does recover from segfaults :)

Results: 165 failed, 132 passed

**Comment by jpienaar** [comment]
What is the delta from before? (and in particular a test that used to pass and now fails)

**Comment by ramiro050** [comment]
> What is the delta from before? (and in particular a test that used to pass and now fails)

Should've explained a bit more. The 297 tests I am running are all tests that used to fail on the `row-major not supported` error. Will update with the categories of errors in a bit.

**Comment by ramiro050** [comment]
**Error Message:** jaxlib.xla_extension.XlaRuntimeError: INVALID_ARGUMENT: external/iree_core/runtime/src/iree/modules/hal/utils/buffer_diagnostics.c:191: INVALID_ARGUMENT; input 0 element type mismatch; expected f32 (21000020) but have bf16 (22000010); while ...

**Failed Tests:**
```
jax/tests/lax_test.py::LaxTest::testConvGeneralDilatedLocal0
jax/tests/lax_test.py::LaxTest::testConvGeneralDilatedLocal1
jax/tests/lax_test.py::LaxTest::testConvGeneralDilatedLocal4
jax/tests/lax_test.py::LaxTest::testConvGeneralDilatedLocal6
jax/tests/lax_test.py::LaxTest::testConvGeneralDilatedLocal8
jax/tests/lax_test.py::LaxTest::testConvTranspose0D1
jax/tests/lax_test.py::LaxTest::testConvTranspose0D2
jax/tests/lax_test.py::LaxTest::testConvTranspose0D5
jax/tests/lax_test.py::LaxTest::testConvTranspose0D8
jax/tests/lax_test.py::LaxTest::testConvTranspose1D1
jax/tests/lax_test.py::LaxTest::testConvTranspose1D3
jax/tests/lax_test.py::LaxTest::testConvTranspose1D4
jax/tests/lax_test.py::LaxTest::testConvTranspose1D5
jax/tests/lax_test.py::LaxTest::testConvTranspose1D6
jax/tests/lax_test.py::LaxTest::testConvTranspose1D7
jax/tests/lax_test.py::LaxTest::testConvTranspose2D6
jax/tests/lax_test.py::LaxTest::testConvTranspose2DT6
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp155
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp157
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp341
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp342
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp472
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp547
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp557
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp567
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp577
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp583
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp588
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp633
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp638
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp643
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp648
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp653
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp658
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp663
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp668
jax/tests/lax_vmap_test.py::LaxVmapTest::testBroadcast5
jax/tests/lax_vmap_test.py::LaxVmapTest::testClamp2
jax/tests/lax_vmap_test.py::LaxVmapTest::testGather7
jax/tests/lax_vmap_test.py::LaxVmapTest::testReshape2
jax/tests/lax_vmap_test.py::LaxVmapTest::testScatterAdd1
jax/tests/lax_vmap_test.py::LaxVmapTest::testScatterAdd2
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelect4
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndGatherAdd5
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndGatherAdd6
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndGatherAdd9
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndScatterAdd2
jax/tests/lax_vmap_test.py::LaxVmapTest::testVariadicReduce8
```
**Error Message:** jaxlib.xla_extension.XlaRuntimeError: INVALID_ARGUMENT: iree/integrations/pjrt/common/api_impl.cc:1184: INVALID_ARGUMENT; :

This is an error during compilation: https://github.com/openxla/openxla-pjrt-plugin/blob/a0f1910b1f7f5ab39f5961aaebabdad40454cf35/iree/integrations/pjrt/common/api_impl.cc#L1159

**Failed Tests:**
```
jax/tests/custom_linear_solve_test.py::CustomLinearSolveTest::test_custom_linear_solve_cholesky
jax/tests/custom_root_test.py::CustomRootTest::test_custom_root_with_aux
jax/tests/custom_root_test.py::CustomRootTest::test_custom_root_with_custom_linear_solve
jax/tests/linalg_test.py::NumpyLinalgTest::testCholesky0
jax/tests/linalg_test.py::NumpyLinalgTest::testCholesky1
jax/tests/linalg_test.py::NumpyLinalgTest::testCholesky4
jax/tests/linalg_test.py::NumpyLinalgTest::testCholesky5
jax/tests/linalg_test.py::NumpyLinalgTest::testCholesky8
jax/tests/linalg_test.py::NumpyLinalgTest::testCholesky9
```
**Error Message:** AssertionError:

Haven't checked all of them, but assertion error usually means output tensor did not match expected tensor.

**Failed Tests:**
```
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=0_concat=1_vmap=2
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=0_concat=1_vmap=3
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=0_concat=2_vmap=2
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=0_concat=2_vmap=3
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=1_concat=0_vmap=2
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=1_concat=0_vmap=3
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=1_concat=2_vmap=2
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=1_concat=2_vmap=3
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=2_concat=0_vmap=2
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=2_concat=0_vmap=3
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=2_concat=1_vmap=2
jax/tests/batching_test.py::BatchingTest::testAllToAll_split=2_concat=1_vmap=3
jax/tests/batching_test.py::BatchingTest::testConcatenate
jax/tests/lax_numpy_test.py::LaxBackedNumpyTests::testArgpartition2
jax/tests/lax_numpy_test.py::LaxBackedNumpyTests::testArgpartition9
jax/tests/lax_test.py::LaxTest::testConvGeneralDilatedLocal7
jax/tests/lax_test.py::LaxTest::testConvGeneralDilatedLocal9
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp357
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp367
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp447
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp487
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp517
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp540
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp549
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp550
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp559
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp560
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp570
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp579
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp587
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp637
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp647
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp657
jax/tests/lax_vmap_op_test.py::LaxVmapOpTest::testOp667
jax/tests/lax_vmap_test.py::LaxVmapTest::testClamp0
jax/tests/lax_vmap_test.py::LaxVmapTest::testClamp3
jax/tests/lax_vmap_test.py::LaxVmapTest::testClamp5
jax/tests/lax_vmap_test.py::LaxVmapTest::testClamp6
jax/tests/lax_vmap_test.py::LaxVmapTest::testClamp7
jax/tests/lax_vmap_test.py::LaxVmapTest::testClamp8
jax/tests/lax_vmap_test.py::LaxVmapTest::testClamp9
jax/tests/lax_vmap_test.py::LaxVmapTest::testGather5
jax/tests/lax_vmap_test.py::LaxVmapTest::testGather9
jax/tests/lax_vmap_test.py::LaxVmapTest::testReduce3
jax/tests/lax_vmap_test.py::LaxVmapTest::testReduce6
jax/tests/lax_vmap_test.py::LaxVmapTest::testReshape1
jax/tests/lax_vmap_test.py::LaxVmapTest::testReshape4
jax/tests/lax_vmap_test.py::LaxVmapTest::testReshape5
jax/tests/lax_vmap_test.py::LaxVmapTest::testReshape7
jax/tests/lax_vmap_test.py::LaxVmapTest::testScatterAdd0
jax/tests/lax_vmap_test.py::LaxVmapTest::testScatterAdd4
jax/tests/lax_vmap_test.py::LaxVmapTest::testScatterAdd5
jax/tests/lax_vmap_test.py::LaxVmapTest::testScatterAdd6
jax/tests/lax_vmap_test.py::LaxVmapTest::testScatterAdd8
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndGatherAdd0
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndGatherAdd1
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndGatherAdd2
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndGatherAdd3
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndGatherAdd4
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndGatherAdd7
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndGatherAdd8
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndScatterAdd4
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndScatterAdd5
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndScatterAdd6
jax/tests/lax_vmap_test.py::LaxVmapTest::testSelectAndScatterAdd8
jax/tests/lax_vmap_test.py::LaxVmapTest::testSqueeze1
jax/tests/lax_vmap_test.py::LaxVmapTest::testSqueeze6
jax/tests/lax_vmap_test.py::LaxVmapTest::testSqueeze7
jax/tests/lax_vmap_test.py::LaxVmapTest::testVariadicReduce0
jax/tests/lax_vmap_test.py::LaxVmapTest::testVariadicReduce1
jax/tests/lax_vmap_test.py::LaxVmapTest::testVariadicReduce3
jax/tests/lax_vmap_test.py::LaxVmapTest::testVariadicReduce4
jax/tests/lax_vmap_test.py::LaxVmapTest::testVariadicReduce5
jax/tests/lax_vmap_test.py::LaxVmapTest::testVariadicReduce6
jax/tests/lax_vmap_test.py::LaxVmapTest::testVariadicReduce9
jax/tests/sparse_test.py::SparseObjectTest::test_dense_round_trip12
jax/tests/sparse_test.py::SparseObjectTest::test_dense_round_trip14
jax/tests/sparse_test.py::SparseObjectTest::test_dense_round_trip8
jax/tests/sparse_test.py::SparseObjectTest::test_todense_CSC
jax/tests/sparse_test.py::SparseObjectTest::test_transpose10
jax/tests/sparse_test.py::SparseObjectTest::test_transpose12
jax/tests/sparse_test.py::SparseObjectTest::test_transpose14
jax/tests/sparse_test.py::SparseObjectTest::test_transpose8
jax/tests/state_test.py::StatePrimitivesTest::test_vmap1
jax/tests/state_test.py::StatePrimitivesTest::test_vmap3
jax/tests/state_test.py::StatePrimitivesTest::test_vmap5
jax/tests/state_test.py::StatePrimitivesTest::test_vmap9
```
**Error Message:** SIGSEGV

**Failed Tests:**
```
jax/tests/lax_test.py::LaxTest::testConvGeneralDilatedLocal2
jax/tests/lax_vmap_test.py::LaxVmapTest::testScatterAdd3
jax/tests/sparse_test.py::SparseObjectTest::test_dense_round_trip10
jax/tests/sparse_test.py::SparseObjectTest::test_dense_round_trip11
jax/tests/sparse_test.py::SparseObjectTest::test_dense_round_trip13
jax/tests/sparse_test.py::SparseObjectTest::test_dense_round_trip15
jax/tests/sparse_test.py::SparseObjectTest::test_dense_round_trip9
jax/tests/sparse_test.py::SparseObjectTest::test_transpose11
jax/tests/sparse_test.py::SparseObjectTest::test_transpose13
jax/tests/sparse_test.py::SparseObjectTest::test_transpose15
jax/tests/sparse_test.py::SparseObjectTest::test_transpose9
```
**Error Message:** jaxlib.xla_extension.XlaRuntimeError: PERMISSION_DENIED: external/iree_core/runtime/src/iree/hal/buffer.c:362: PERMISSION_DENIED; requested usage was not specified when the buffer was allocated; buffer allows DISPATCH_INDIRECT_PARAMS|DISPATC...

**Failed Tests:**
```
jax/tests/lax_vmap_test.py::LaxVmapTest::testScatterAdd9
```
**Error Message:** NotImplementedError: MLIR translation rule for primitive 'eigh' not found for platform iree_cpu

**Failed Tests:**
```
jax/tests/svd_test.py::SvdTest::testSvdWithRectangularInput0
jax/tests/svd_test.py::SvdTest::testSvdWithRectangularInput1
jax/tests/svd_test.py::SvdTest::testSvdWithRectangularInput2
jax/tests/svd_test.py::SvdTest::testSvdWithRectangularInput3
jax/tests/svd_test.py::SvdTest::testSvdWithRectangularInput4
jax/tests/svd_test.py::SvdTest::testSvdWithRectangularInput5
jax/tests/svd_test.py::SvdTest::testSvdWithRectangularInput6
jax/tests/svd_test.py::SvdTest::testSvdWithRectangularInput7
jax/tests/svd_test.py::SvdTest::testSvdWithRectangularInput8
jax/tests/svd_test.py::SvdTest::testSvdWithRectangularInput9
```

**Comment by ramiro050** [comment]
The only failing tests that could be related to this implementation are the `AssertionError` ones and the crashes: 98 in total.

**Comment by ramiro050** [comment]
Updated the list of failures after rebasing to latest main. The rebase fixed the bf16 errors. We are now at: 126 failed, 171 passed

**Comment by jpienaar** [comment]
> Haven't checked all of them, but assertion error usually means output tensor did not match expected tensor.

I wonder if we need another transpose in the copy back from the device too for these cases.

**Comment by ramiro050** [comment]
The way I was constructing the `input_strides` was incorrect. After fixing, we are at: 54 failed, 243 passed. I've updates the list of failing tests above.

**Review by jpienaar — APPROVED** [review]

**Pull request merged**